Backport ccm main

Lorak-mmk opened the pull request:
This is a backport of the CCM test support developer during hackathon onto main branch.

I mostly took the current state from branch-hackathon. One important change I made is to remove separate `ccm-integration` test target. Instead ccm tests are now a module in `integration`.
This avoids the problem with sharing utils. It should also be quicker to compile - no need to link 2 separate binaries.

logged_cmd and ip_allocator were copied with very little changes. In logged_cmd I added a support for non-async command running, needed for destroy_sync in CCM API.
I split CCM API code into two parts:
- `cli_wrapper` module, which contains a thin layer over CCM cli. It is meant only as  way to translate CCM flags to Rust code.
- Previously existing cluster.rs, which I divided into cluster.rs and node.rs. They contain API to be used by tests.

This code is fully internal to the crate, so we can change it freely, thus there is no need to spend too much time on it - we can always improve later.

Big parts of the test-API code are unused. This is intentional - I don't want to introduce more CCM tests than necessary, because it would increase CI time.
This code will become used when we introduce other tests.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- [ ] I added appropriate `Fixes:` annotations to PR description.

Lorak-mmk commented:
Possible improvements / API changes after a brief glance at the code:
- Should `NodeStartOptions` be a struct? In other words, does it make sense to enable e.g. `no_wait` and `wait_other_notice` at the same time? We need to know the exact semantics of the wait-related flags to know that (cc @fruch because I don't think this is documented anywhere in this cursed software).
- Same for NodeStopOptions
- We hold nodes in Arc<RwLock<>>, and methods that give user nodes return that. Maybe we could return refs / mut refs and get rid of Arc<RwLock<>>? I'm not sure.
- In the future when we have custom test runner, we could make new struct (`ClusterPreferences` + `NodePreferences`) or extend `*Options` structs. Why? Tests may not care about some parameters, which could help test runner to provide less clusters.

muzarski commented:
> On the above matter @muzarski : How should I adapt CCM given that we now support multiple TLS backends? I see we have in mod.rs DB_TLS_CERT_PATH, DB_TLS_KEY_PATH and CA_TLS_CERT_PATH (which I removed for now because it was guarded by old feature name). I also see that on branch-hackathon you wrote a TLS test. We should probably make a test per backend, right? What about those vars? Do they make sense for all the tests? In that case we just need to change feature guard on CA_TLS_CERT_PATH to be activated when any backend is active - or to even make it always active because why not.

Notice that I implemented all of this when we had old certificates in the repository. We had to update them, because of the errors thrown by `rustls`.

### Why old certs worked for `openssl` but did not for `rustls`?
It's because `rustls` supports hostname verification **by default**, while openssl does not. The `CN` (common name) in db certificate was not matching the hostname, thus `rustls` was throwing an error.

### What changed in the certificates, compared to the previous version?
I generated the db certificates assigned to static IP (172.44.0.2 - one we currently use in CI for TLS single-node cluster). In other words, the extensions to certificate request in openssl config looked like:
```
[v3_req]
# Extensions to add to a certificate request
basicConstraints = CA:FALSE
subjectAltName = IP:172.44.0.2
keyUsage = nonRepudiation, digitalSignature, keyEncipherment
```

Thanks to that, `rustls` is able to verify the hostname using node's IP. It checks whether the node that we try to connect to has the same IP as the one defined in certificate (under `subjectAltName`).

### Current state
Currently, our SSL "tests" are limited to simply running the `tls-openssl` and `tls-rustls` examples in SSL CI workflow.
On hackathon-branch, however, I removed the SSL workflow and migrated the example test to ccm. It obviously did it for openssl only, as rustls was not supported back then.

### Implementing the corresponding ccm test for rustls backed and removing SSL CI workflow.
Well, this is a bit tricky. While, it was not an issue with `openssl`, for the reasons stated above (no hostname verification), it won't work for `rustls`. This is because we use dynamic IPs in ccm tests.

The temporary solution I see: migrate the tests from SSL workflow to ccm only for `openssl` and limit rustls "tests" to just running the example against the global cluster (as it is currently done in CI).

If we ever decide that we want to have ccm tests for rustls backend, I list the possible solutions to the dynamic IP problem:
- Generate certificates on the fly, during the test. We would firstly receive some IP from the `IpAllocator` (or some other mechanism in the future) and then we could generate the self-signed cert for this IP. Then we could `ccm updateconf` and provide the path to generated certificate. There are some crates we could use, e.g. https://docs.rs/rcgen/latest/rcgen/.
- Disable hostname verification during the tests. I think this can be configured via the trait: [`rustls::client::danger::ServerCertVerifier`](https://docs.rs/rustls/latest/rustls/client/danger/trait.ServerCertVerifier.html). I'm not entirely sure, though. This needs some research - I think this is placed in `danger` module for a reason. OTOH, we already use it in `scylla::cloud::config` and implement it for our `NoCertificateVerification` struct.

fruch commented:
> Moving such workflows to CCM has 2 advantages:
> 
> * We get rid of custom images
> * We get rid of a GHA workflow
> 
> If we move only part of it, we get neither. So I'll cherry pick commits that move auth, and skip TLS for now. We can do that in the future.
> 
> Btw is it possible to use domain names instead of ip addresses with scylla? In other words, can we have domain names instead of ip addresses in system.peers in driver-relevant columns? If it was possible, we could use certs with hostnames instead of ips.

scylla can use hostnames, but then you need a dns server to map them.

I think generating certs as needed is the best approach, and also give the flexibility to try more variants as needed.
that's what we are doing in dtest, and in SCT.

Lorak-mmk commented:
> 
> scylla can use hostnames, but then you need a dns server to map them.
> 
> I think generating certs as needed is the best approach, and also give the flexibility to try more variants as needed. that's what we are doing in dtest, and in SCT.

Is there functionality in CCM to generate certs? Or do we have to do it other way?

If Scylla can use hostnames, then we should test it too.


@fruch one other question for you. Could you describe (or point to documentation if such exists) what exactly wait-related flags do in CCM, and how do they interact if I specify more than one?

fruch commented:
> > scylla can use hostnames, but then you need a dns server to map them.
> > I think generating certs as needed is the best approach, and also give the flexibility to try more variants as needed. that's what we are doing in dtest, and in SCT.
> 
> Is there functionality in CCM to generate certs? Or do we have to do it other way?
> 
> If Scylla can use hostnames, then we should test it too.
> 
> @fruch one other question for you. Could you describe (or point to documentation if such exists) what exactly wait-related flags do in CCM, and how do they interact if I specify more than one?

you are more then welcome to document it.

Lorak-mmk commented:
> you are more then welcome to document it.

I'd be happy to make a PR that improves descriptions, but I would have to first understand those options myself.
I don't know ccm's codebase at all, and it is not really friendly to new contributors, that's why I asked you to explain those options.

Lorak-mmk commented:
I backported the commits that move auth to CCM. I also removed TLS support from CCM for now.

Lorak-mmk commented:
Marking as ready. The way I see it the only improvement I can make here is better CCM API - which needs input from others, which is basically a review.

muzarski reviewed (changes requested):
> We hold nodes in Arc<RwLock<>>, and methods that give user nodes return that. Maybe we could return refs / mut refs and get rid of Arc<RwLock<>>? I'm not sure.

As of now, there is no use case for `Arc<RwLock<>>` - I think we can return refs/mutrefs for now. We could always revert this in the future. It also simplifies the API - I believe `NodeList` is no longer necessary then. Instead, we can expose `nodes_iter_[mut]()` and `get_node_[mut]_by_id` methods on `Cluster`.

Currently, `append_node()` and `add_node()` methods return `Arc<RwLock<>>`. They could return node id instead.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/ip_allocator.rs:
```diff
@@ -0,0 +1,230 @@
+    }
+}
+
+impl From<NetPrefix> for String {
+    fn from(value: NetPrefix) -> Self {
+        value.0.to_string()
+    }
+}
```
Nothing broke when I removed it, so I'll keep it removed.

Lorak-mmk commented:
- Addressed @muzarski 's comments
- Removed all the Arc<Mutex<Node>> stuff, now we just operate on Nodes.
- I decided to retain NodeList because I welcome any kind of separation and structure in this code. I made its method simpler using iterator methods.

Lorak-mmk commented:
I have one more idea: we can split off another file from cluster.rs, I would call it `ccm_cmd.rs`.
It would be a simple wrapper over CCM, providing builder-style commands.
The purpose of this module would be to provide convenient way to call CCM, and encode all its commands and flags into Rust types.
`cluster.rs` would be responsible for providing user-facing API, handling config dirs etc. Its code would hopefully become cleaner.

Lorak-mmk commented:
I did this for 2 commands as an experiment, in additional commit. I like the new version, so unless anyone has different opinion I'll convert the rest of the command to this.

@dkropachev I see that both `ccm create` and `ccm populate` accept `ipprefix` argument. Why? What are their respective semantics?

dkropachev commented:
> I did this for 2 commands as an experiment, in additional commit. I like the new version, so unless anyone has different opinion I'll convert the rest of the command to this.
> 
> @dkropachev I see that both `ccm create` and `ccm populate` accept `ipprefix` argument. Why? What are their respective semantics?

`ccm create` calls `ccm populate`, if you provide number of nodes.
Flow of `ccm create` + `ccm add` + `ccm start` is broken, so you better make `ccm create` to call `ccm populate` to create a cluster in the single blow.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/ccm_cmd.rs:
```diff
@@ -0,0 +1,123 @@
+
+impl Ccm {
+    pub(crate) fn cluster_create(
+        &mut self,
+        name: String,
+        version: String,
+        ip_prefix: NetPrefix,
+        db_type: DBType,
+    ) -> ClusterCreate<'_> {
+        ClusterCreate {
+            ccm: self,
+            name,
+            version,
+            ip_prefix: ip_prefix.to_string(),
+            db_type,
+        }
+    }
+}
+
+impl ClusterCreate<'_> {
+    pub(crate) async fn run(self) -> Result<ExitStatus, anyhow::Error> {
+        let mut args: Vec<&str> = vec![
+            "create",
+            self.name.as_str(),
+            "--version",
+            self.version.as_str(),
+            "--ipprefix",
+            self.ip_prefix.as_str(),
+            "--config-dir",
+            self.ccm.config_dir.as_str(),
+        ];
+        if self.db_type == DBType::Scylla {
+            args.push("--scylla");
+        }
+        self.ccm
+            .cmd
+            .run_command("ccm", &args, RunOptions::new())
+            .await
+    }
+}
```
Right now all parameters are accepted in the constructor of the command. I'll switch it to builder pattern soon, and of course move other commands too.

Lorak-mmk commented:
New version of the PR. Finalized the move to a separate command builders and made a lot of other changes.
I think there is a lot of room for improvement still, but I'd like to put this up for review now anyway - it will be easier to get it to something acceptable together.

muzarski commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,362 @@
+        Ok(())
+    }
+
+    fn append_node(&mut self, node_options: NodeOptions) -> &mut Node {
+        let node_name = node_options.name();
+        let node = Node::new(node_options, self.ccm_cmd.for_node(node_name));
+
+        self.nodes.push(node);
+        self.nodes.0.last_mut().unwrap()
+    }
```
I like that we return `&mut Node` here. (previously that was `Arc<RwLock<_>>`, correct?)

Lorak-mmk replied:
I'm not sure tbh.

muzarski reviewed (approved):
Looks much better.

Lorak-mmk commented:
Addressed Mikolaj's comment.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cli_wrapper/cluster.rs:
```diff
@@ -0,0 +1,330 @@
+use super::node::NodeCcm;
+use super::{DBType, NodeStartOptions};
+
+pub(crate) struct Ccm {
+    config_dir: String,
+    cmd: Arc<LoggedCmd>,
+}
+
+impl Ccm {
+    pub(crate) fn new(config_dir: impl AsRef<Path>, cmd: Arc<LoggedCmd>) -> Self {
+        Self {
+            config_dir: config_dir
+                .as_ref()
+                .to_str()
+                .expect("Config dir not representable as str")
+                .to_string(),
+            cmd,
+        }
+    }
```
I do need to perform the conversion at some point in order to put the config dir in argument list.
Imo it is better to do it earlier (and fail earlier, and do the conversion only once) instead of doing it in every command, duplicating code.

wprzytula replied:
If putting in the argument list requires conversion to `String`, then I agree. But maybe `OsString` is enough there? Because IIUC, `PathBuf`, is built atop `OsString`.

Lorak-mmk replied:
It may be enough when adding an argument to Tokio / std `Command` - but note that we pass an array / vector of arguments. What should be the type for such array? Currently I'm using String or &str, depending on command.
Switching to OsString will be ugly for all other arguments apart from config path.

wprzytula replied:
Agreed. Then disregard my suggestion.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,355 @@
+                    )?;
+                    }
+                    _ => {
+                        return Err(Error::from(err).context(format!(
+                            "failed to create root directory {:?}",
+                            config_dir_path
+                        )));
```
As far as I can tell: there are 2 `context` methods in anyhow:
- In `Context` trait: https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.context . This trait is implemented for `Result` from std, but not for `std::io::Error` 
- On `anyhow::Error`: https://docs.rs/anyhow/latest/anyhow/struct.Error.html#method.context - which this code creates.

If you see some more idiomatic way to handle this please let me know because I'm also not a fan of this code.

wprzytula replied:
As a side note, it's also present on `Option`.

wprzytula replied:
> If you see some more idiomatic way to handle this please let me know because I'm also not a fan of this code.

Unfortunately, I don't see a better way.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,355 @@
+    }
+
+    // Only for use in desctructor - that is why it takes `&mut`.
+    // This code is here, instead of directly in destructor, because
+    // its logic is coupled with `destroy`. Putting those methods
+    // close decreases the chances of forgetting to change one of them
+    // in the future.
+    fn destroy_sync(&mut self) -> Result<(), Error> {
+        if !self.destroyed {
+            self.ccm_cmd.cluster_remove().run_sync().map(|_| ())?;
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn destroy(mut self) -> Result<(), Error> {
+        self.stop().await?;
+        self.ccm_cmd.cluster_remove().run().await.map(|_| ())?;
+        self.destroyed = true;
+        Ok(())
+    }
```
I actually don't think we need to run `stop()` anywhere - afaik `ccm remove` stops the cluster. @dkropachev please confirm.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,355 @@
+
+pub(crate) struct NodeList(Vec<Node>);
+
+#[allow(dead_code)]
+impl NodeList {
+    fn get_free_node_id(&self) -> Option<NodeId> {
+        for node_id in 1..=255 {
+            match self.iter().find(|node| node.id() == node_id) {
+                Some(_) => continue,
+                None => return Some(node_id),
+            }
+        }
+        None
+    }
```
I made it more granular (not yet pushed). There is a lot of unused stuff.
In principle we could write tests (or extend the example test) in order to not make it unused, but I'm not a fan of it. Doing more ccm operations means longer running tests, so I prefer to just keep this unused stuff, and wait for new tests to start using it.
After some time, if something remains truly unused we can remove it.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,355 @@
+        self.nodes.0.last_mut().unwrap()
+    }
+
+    pub(crate) async fn add_node(
+        &mut self,
+        datacenter_id: Option<u16>,
+    ) -> Result<&mut Node, Error> {
+        let id = self.nodes.get_free_node_id().expect("No available node id");
+        let datacenter_id = datacenter_id.unwrap_or(1);
+        let node_options = NodeOptions {
+            id,
+            datacenter_id,
+            ..NodeOptions::from_cluster_opts(&self.opts)
+        };
```
Let's add such possibility when the need arises. I don't think it is the purpose here to anticipate all possible uses.

wprzytula replied:
I'm convinced that testing non-uniform clusters (wrt number of shards, especially) is important. I recall that such scenarios have shown to be handled poorly in some edge cases in the past.
If we don't add such option, we'll discourage test writers (possibly ourselves) from creating such test scenarios.

Lorak-mmk replied:
> However, I assume you had grasped it before you set this PR as ready for review.

The same could be said about any functionality. The purpose of this PR is to add minimal support. It can be extended later.

> I recall that such scenarios have shown to be handled poorly in some edge cases in the past.

I remember it being handled poorly by Scylla, not by Driver. Anyway, even if this is true it does not imply it needs to be part of initial minimal support.

Lorak-mmk replied:
https://github.com/scylladb/scylla-rust-driver/issues/1367

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cli_wrapper/cluster.rs:
```diff
@@ -0,0 +1,330 @@
+}
+
+impl ClusterStart<'_> {
+    pub(crate) fn wait_options(mut self, options: Option<NodeStartOptions>) -> Self {
+        self.wait_opts = options;
+        self
+    }
+    pub(crate) fn scylla_smp(mut self, smp: u16) -> Self {
+        self.scylla_smp = Some(smp);
+        self
+    }
+
+    pub(crate) fn scylla_mem_megabytes(mut self, mem_megabytes: u32) -> Self {
+        self.scylla_mem_megabytes = Some(mem_megabytes);
+        self
+    }
```
The way I understand it being able to use `None` in the public interface (like `Cluster::start`) was a usability request from a hackathon participant, which I agree with.
Now comes the question: what to use in internal interface (cli_wrapper) that you commented on?
I chose to also use Option in order to have only a single place to handle default value (in `run` method).

I'm not strongly attached to this, so I can change it if you disagree.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,355 @@
+        self.updateconf(args).await
+    }
+
+    /// This method starts the cluster. User can provide optional [`NodeStartOptions`] to control the behavior of the nodes start.
+    /// If `None` is provided, the default options are used (see the implementation of Default for [`NodeStartOptions`]).
+    pub(crate) async fn start(&mut self, opts: Option<NodeStartOptions>) -> Result<(), Error> {
+        self.ccm_cmd
+            .cluster_start()
+            .wait_options(opts)
+            .scylla_smp(self.opts.smp)
+            .scylla_mem_megabytes(self.opts.memory)
+            .run()
+            .await?;
+
+        for node in self.nodes.iter_mut() {
+            node.set_status(NodeStatus::Started);
+        }
+
+        Ok(())
+    }
+
+    fn append_node(&mut self, node_options: NodeOptions) -> &mut Node {
+        let node_name = node_options.name();
+        let node = Node::new(node_options, self.ccm_cmd.for_node(node_name));
+
+        self.nodes.push(node);
+        self.nodes.0.last_mut().unwrap()
+    }
```
> IIUC, ccm starts with a configurable number of uniform nodes, and then one can add more nodes, not necessarily uniform?

That's how I understand it too.

Lorak-mmk commented on scylla/tests/integration/ccm/lib/cluster.rs:
```diff
@@ -0,0 +1,355 @@
+        self.nodes.0.last_mut().unwrap()
+    }
+
+    pub(crate) async fn add_node(
+        &mut self,
+        datacenter_id: Option<u16>,
+    ) -> Result<&mut Node, Error> {
+        let id = self.nodes.get_free_node_id().expect("No available node id");
+        let datacenter_id = datacenter_id.unwrap_or(1);
```
Ease of use - no need to think about what number to pass in scenarios where you don't care about DCs.

Lorak-mmk commented on Makefile:
```diff
@@ -99,3 +99,16 @@ shell:
 	cargo clean
 	rm -rf docs/book
+
+.PHONY: use_cargo_lock_msrv
+use_cargo_lock_msrv:
+	mv Cargo.lock Cargo.lock.bak
+	mv Cargo.lock.msrv Cargo.lock
+
+.PHONY: restore_cargo_lock
+restore_cargo_lock:
+	mv Cargo.lock Cargo.lock.msrv
+	mv Cargo.lock.bak Cargo.lock
+
+.PHONY: test_cargo_lock_msrv
+test_cargo_lock_msrv: use_cargo_lock_msrv check restore_cargo_lock
```
I'll remove this before merging. If we add it, it should be in a separate PR.

wprzytula replied:
Can you remove this now? I don't see a point in having this in this PR atm if you don't want it to be merged as part of this PR.

Lorak-mmk commented:
Addressed most @wprzytula comments, and responded to rest.
Rebased on main.

wprzytula commented on scylla/tests/integration/ccm/lib/logged_cmd.rs:
```diff
@@ -0,0 +1,494 @@
+                Ok(status)
+            }
+            Err(e) => {
+                tracing::info!(
+                    "{:15} -> failed to wait on child process: = {}",
+                    format!("exited[{}]", run_id),
+                    e
+                );
```
> I suppose it doesn't allocate?

Correct.

wprzytula commented on scylla/tests/integration/ccm/lib/ip_allocator.rs:
```diff
@@ -0,0 +1,224 @@
+    }
+}
+
+/// A local subnet identifier (127.x.y.0/24).
+/// The local subnet is identified by two octets x and y.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+struct LocalSubnetIdentifier(u8, u8);
```
It's understandable that it's not you who wrote the code. However, I assume you had grasped it before you set this PR as ready for review. And while trying to grasp it, I guess you had similar trouble as I had. That's why I would assume you'd have written comments if you had found it hard to understand. But you didn't write them, so I reasoned that the code was clear to you. That's why I informed you here that it's been unclear to me.

wprzytula commented:
Please update the cover letter.

wprzytula reviewed (approved):
LGTM!
